File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/actors/PlayActor.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors;

import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.actors.ZLinkActorContext;
import systems.zlink.framework.actors.ZLinkActorJoinCompletion;
import systems.zlink.framework.actors.ZLinkActorJoinOperationId;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameFailedNotify;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerInfo;
import systems.zlink.samples.tictactoe.shared.contracts.TicTacToeGameJoinRes;

public final class PlayActor implements ZLinkActor {
    private final String actorId;
    private final ZLinkActorContext context;
    private PlayerInfo player;
    private String joinedRoomId;
    private String pendingRoomId;
    private final java.util.Set<ZLinkActorJoinOperationId>
        completedJoinOperations = new java.util.HashSet<>();
    private boolean destroyAfterEntrySpotJoin;
    private boolean disconnected;

    public PlayActor(String actorId, ZLinkActorContext context) {
        this.actorId = actorId;
        this.context = context;
    }

    public String actorId() {
        return actorId;
    }

    @Override
    public ZLinkActorContext context() {
        return context;
    }

    public void applyPlayer(PlayerInfo player) {
        if (!actorId.equals(player.actorId())) {
            throw new IllegalArgumentException("player actor id does not match actor");
        }
        this.player = player;
    }

    public PlayerInfo requirePlayer() {
        if (player == null) {
            throw new IllegalStateException("actor has not been authenticated");
        }
        return player;
    }

    public PlayerInfo playerOrNull() {
        return player;
    }

    public int incrementWins() {
        PlayerInfo current = requirePlayer();
        PlayerInfo updated = new PlayerInfo(
            current.actorId(),
            current.displayName(),
            current.level(),
            current.wins() + 1);
        this.player = updated;
        return updated.wins();
    }

    public void joinGame(String roomId) {
        this.joinedRoomId = roomId;
    }

    public void trackDeferredJoin(String roomId) {
        if (pendingRoomId != null) {
            throw new IllegalStateException("a room join is already pending");
        }
        pendingRoomId = roomId;
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> onJoinCompleted(
        ZLinkActorJoinCompletion completion) {
        ZLinkActorJoinOperationId operationId = completion
            instanceof ZLinkActorJoinCompletion.Accepted accepted
                ? accepted.operationId()
                : completion instanceof ZLinkActorJoinCompletion.Rejected rejected
                    ? rejected.operationId()
                    : ((ZLinkActorJoinCompletion.Failed) completion).operationId();
        if (!completedJoinOperations.add(operationId)) {
            return java.util.concurrent.CompletableFuture.completedFuture(null);
        }
        String roomId = pendingRoomId;
        pendingRoomId = null;
        if (completion instanceof ZLinkActorJoinCompletion.Accepted accepted) {
            TicTacToeGameJoinRes reply = accepted.reply()
                .decode(TicTacToeGameJoinRes.class);
            if (roomId == null || roomId.isBlank()) {
                // A remote join completes on the relocated Actor. The target
                // reconstructs the room from the durable accepted reply.
                roomId = reply.state().roomId();
            }
            joinGame(roomId);
            return context.boundSession()
                .send(new JoinGameRes(reply.state()))
                .submit();
        }
        if (roomId == null) {
            roomId = "";
        }
        if (completion instanceof ZLinkActorJoinCompletion.Rejected) {
            return context.boundSession()
                .send(new JoinGameFailedNotify(roomId, "Rejected", false))
                .submit();
        }
        var failed = (ZLinkActorJoinCompletion.Failed) completion;
        return context.boundSession()
            .send(new JoinGameFailedNotify(
                roomId, failed.kind().name(), false))
            .submit();
    }

    public String requireJoinedGame() {
        if (joinedRoomId == null || joinedRoomId.isEmpty()) {
            throw new IllegalStateException("actor has not joined a room");
        }
        return joinedRoomId;
    }

    public String joinedRoomId() {
        return joinedRoomId;
    }

    public boolean destroyAfterEntrySpotJoin() {
        return destroyAfterEntrySpotJoin;
    }

    public void markForDestroyAfterRoomLeave() {
        destroyAfterEntrySpotJoin = true;
    }

    public boolean disconnected() {
        return disconnected;
    }

    public void markDisconnected() {
        disconnected = true;
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/actors/PlayActor.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors

import systems.zlink.framework.actors.ZLinkActorContext
import systems.zlink.framework.actors.ZLinkActorJoinCompletion
import systems.zlink.framework.actors.ZLinkActorJoinOperationId
import systems.zlink.framework.kotlin.ZLinkSuspendingActor
import systems.zlink.framework.kotlin.decode
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameFailedNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerInfo
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.TicTacToeGameJoinRes

class PlayActor(val actorId: String, override val context: ZLinkActorContext) :
    ZLinkSuspendingActor() {
    private var joinedRoomId: String? = null
    private var pendingRoomId: String? = null
    private val completedJoinOperations = mutableSetOf<ZLinkActorJoinOperationId>()
    private var player: PlayerInfo? = null
    var destroyAfterEntrySpotJoin: Boolean = false
        private set

    var disconnected: Boolean = false
        private set

    fun applyPlayer(player: PlayerInfo) {
        require(player.actorId == actorId) { "player actor id does not match actor" }
        this.player = player
    }

    fun requirePlayer(): PlayerInfo =
        player ?: throw IllegalStateException("actor has not been authenticated")

    fun playerOrNull(): PlayerInfo? = player

    fun incrementWins(): Int {
        val current = requirePlayer()
        val updated = current.copy(wins = current.wins + 1)
        player = updated
        return updated.wins
    }

    fun joinGame(roomId: String) {
        joinedRoomId = roomId
    }

    fun trackDeferredJoin(roomId: String) {
        check(pendingRoomId == null) { "a room join is already pending" }
        pendingRoomId = roomId
    }

    // --8<-- [start:doc-join-completed]
    override suspend fun onJoinCompletedSuspending(completion: ZLinkActorJoinCompletion) {
        val operationId =
            when (completion) {
                is ZLinkActorJoinCompletion.Accepted -> completion.operationId()
                is ZLinkActorJoinCompletion.Rejected -> completion.operationId()
                is ZLinkActorJoinCompletion.Failed -> completion.operationId()
            }
        if (!completedJoinOperations.add(operationId)) {
            return
        }

        var roomId = pendingRoomId
        pendingRoomId = null
        when (completion) {
            is ZLinkActorJoinCompletion.Accepted -> {
                val reply = completion.reply().decode<TicTacToeGameJoinRes>()
                if (roomId.isNullOrBlank()) {
                    // The relocated Actor reconstructs the room from the durable accepted reply.
                    roomId = reply.state.roomId
                }
                joinGame(roomId.orEmpty())
                context.boundSession().kotlin().send(JoinGameNotify(reply.state)).await()
            }
            is ZLinkActorJoinCompletion.Rejected ->
                context
                    .boundSession()
                    .kotlin()
                    .send(JoinGameFailedNotify(roomId.orEmpty(), "Rejected"))
                    .await()
            is ZLinkActorJoinCompletion.Failed ->
                context
                    .boundSession()
                    .kotlin()
                    .send(JoinGameFailedNotify(roomId.orEmpty(), completion.kind().name))
                    .await()
        }
    }

    // --8<-- [end:doc-join-completed]

    fun requireJoinedGame(): String =
        joinedRoomId ?: throw IllegalStateException("actor has not joined a game")

    fun joinedRoomIdOrNull(): String? = joinedRoomId

    fun markForDestroyAfterRoomLeave() {
        destroyAfterEntrySpotJoin = true
    }

    fun markDisconnected() {
        disconnected = true
    }
}
